src/Microsoft.Restier.Core/Query/QueryExpressionContext.cs

```diff
@@ -125,9 +126,76 @@ public void PopVisitedNode()
                 return new QueryModelReference(source.EntitySet, source.Type);
             }
 
-            // source is a sequence of T1 and output is a sequence of T2
-            // Like query People(key)/Trips or People/NS.DerivedPeople
-            // TODO Null is return in these cases now, need return correct value
+            // In case sourceType IEnumerable<Person> and resultType is IEnumerable<SelectExpandBinder.SelectAllAndExpand<Person>>
+            // or IEnumerable<SelectExpandBinder.SelectAll<Person>> or IEnumerable<SelectExpandBinder.SelectSome<Person>>
+            // or IEnumerable<SelectExpandBinder.SelectSomeAndInheritance<Person>>
+            if (sourceType != null && resultType != null)
+            {
+                var resultGenericType = resultType.GenericTypeArguments[0];
+                if (resultGenericType.IsGenericType)
+                {
+                    var resultElementType = resultGenericType.GenericTypeArguments[0];
+                    var sourceElementType = sourceType.GenericTypeArguments[0];
+
+                    // Handle source is type of sub class and result is a base class
+                    if (resultElementType == sourceElementType || sourceElementType.IsSubclassOf(resultElementType))
+                    {
+                        return new QueryModelReference(source.EntitySet, source.Type);
+                    }
+                }
+            }
+
+            // In this case, the sourceType is null
+            if (method.Name.Equals("OfType"))
+            {
+                // Did not consider multiple namespaces have same entity type case
+                var entityName = resultType.GenericTypeArguments[0].Name;
+                var emdEntityType = model.SchemaElements.SingleOrDefault(e => e.Name == entityName 
+                    && e.SchemaElementKind == EdmSchemaElementKind.TypeDefinition);
+                var collType = new EdmCollectionType(new EdmEntityTypeReference((IEdmEntityType)emdEntityType, false));
+                return new QueryModelReference(source.EntitySet, collType);
+            }
+
+            // Till here, it means the result is not part of previous result and entity set will be null
+            // This mean result is a collection as resultType is IEnumerable<>
+            if (resultType != null)
+            {
+                // Did not consider multiple namespaces have same entity type case
+                var typeName = resultType.GenericTypeArguments[0].Name;
+                var emdSchemaType = model.SchemaElements.SingleOrDefault(e => e.Name == typeName 
+                    && e.SchemaElementKind == EdmSchemaElementKind.TypeDefinition);
+                
+                // This means result is collection of Entity/Complex/Enum
+                IEdmTypeReference edmTypeReference = null;
```

Comment:
consider find existing extension method from ODL or EDM. Or extract these EDM-related stuff into a separate class containing extension methods to make the code compact.